pkg/frontend/querymiddleware/query_limiter.go

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/blake2b"
+
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+// queryLimiterMiddleware blocks a query if it is should not be executed more frequently than a configured interval,
+// and allows it if enough time has passed since the last time the query was allowed.
+// When blocked, the request is rejected with a "too many requests" error.
+// The query limiter currently only matches exact queries, and does not check against query time, only execution time.
+type queryLimiterMiddleware struct {
+	next                  MetricsQueryHandler
+	cache                 cache.Cache
+	keyGen                CacheKeyGenerator
+	limits                Limits
+	logger                log.Logger
+	blockedQueriesCounter *prometheus.CounterVec
+}
+
+func newQueryLimiterMiddleware(
+	cache cache.Cache,
+	cacheKeyGen CacheKeyGenerator,
+	limits Limits,
+	logger log.Logger,
+	blockedQueriesCounter *prometheus.CounterVec,
+) MetricsQueryMiddleware {
+	return MetricsQueryMiddlewareFunc(func(next MetricsQueryHandler) MetricsQueryHandler {
+		return &queryLimiterMiddleware{
+			next:                  next,
+			cache:                 cache,
+			keyGen:                cacheKeyGen,
+			limits:                limits,
+			logger:                logger,
+			blockedQueriesCounter: blockedQueriesCounter,
+		}
+	})
+}
+
+func (ql *queryLimiterMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return ql.next.Do(ctx, req)
+	}
+
+	key := ql.keyGen.QueryRequestLimiter(ctx, tenant.JoinTenantIDs(tenantIDs), req)
+	hashedKey := maybeHashCacheKey(key)
+	// start at max duration value
+	cacheValue := validation.LimitedQuery{
```

Comment:
nit: calling this `cacheValue` made me think we were still storing something in the cache. You also don't need to specify the fields here since `""` and `0` are the default values.